src/devices/ikea_sparsnas.c

```
@@ -0,0 +1,251 @@
+17: uint8_t battery;       // Battery level, 0-100%
+18: uint16_t CRC;          // 16 bit CRC of bytes 0-17
+
+The packet's integrety can be checked with the 16b CRC at the end of the packet. 
+There are also several other ways to check the integrety of the package. 
+ - (preamble)
+ - CRC
+ - The decrypted sensor ID
+ - the constant bytes at 0, 3 and 4
+
+The decryption, CRC is calculation, value extraction and interpretation is 
+taken from https://github.com/strigeus/sparsnas_decoder and adapted to 
+this application. Many thanks to strigeus!
+
+Most other things are from https://github.com/kodarn/Sparsnas which is an
+amazing repository of the IKEA Sparsnäs. Everything is studied with greay
+detail. Many thanks to kodarn!
+
+*/
+
+
+
+#define IKEA_SPARSNAS_SENSOR_ID 617633
+#define IKEA_SPARSNAS_PULSES_PER_KWH 1000
+
+// No settings required below this point
+
+
+#include "decoder.h"
+#define IKEA_SPARSNAS_MESSAGE_BITLEN 160    // 20 bytes incl 8 bit length, 8 bit address, 128 bits data, and 16 bits of CRC. Excluding preamble and sync word
+#define IKEA_SPARSNAS_MESSAGE_BYTELEN    ((IKEA_SPARSNAS_MESSAGE_BITLEN + 7) / 8)
+
+#define IKEA_SPARSNAS_PREAMBLE_BITLEN 32
+static const uint8_t preamble_pattern[4] = {0xAA, 0xAA, 0xD2, 0x01};
+
+
+// Graciously copied from https://github.com/strigeus/sparsnas_decoder
+static uint16_t ikea_sparsnas_crc16(const uint8_t *data, size_t n) {
+  uint16_t crcReg = 0xffff;
+  size_t i, j;
+  for (j = 0; j < n; j++) {
+    uint8_t crcData = data[j];
+    for (i = 0; i < 8; i++) {
+      if (((crcReg & 0x8000) >> 8) ^ (crcData & 0x80))
+        crcReg = (crcReg << 1) ^ 0x8005;
+      else
+        crcReg = (crcReg << 1);
+      crcData <<= 1;
+    }
+  }
+  return crcReg;
+}
+
+static int ikea_sparsnas_callback(r_device *decoder, bitbuffer_t *bitbuffer)
+{
+
+    if (decoder->verbose > 1)
+        decoder_output_bitbufferf(decoder, bitbuffer, "IKEA Sparsnäs:");
+  
+
+    uint8_t buffer[IKEA_SPARSNAS_MESSAGE_BYTELEN];
+    
```

Declare the variables first and then add an early termination condition based on the amount of bits and the preamble check. Then add the crc check. If that succeeds continue with decoding.

---

I am not entirely sure what you mean here. 

Is it that you want variables declared at the top of the function? 

The other things are done in the order they can be done.
1. Preamble checking and extraction
2. CRC Check (needs extracted packet without preamble)
3. Decryption (No need to decrypt if CRC fails)
4. Additional integrity checks (Needs decrypted packet)
5. Data extraction and interpretation (Needs a decrypted and valid packet)

---

The point is that these checks are done for every fsk signal that is detected. So things that are not matching needs to fail fast with easy checks.

---

The if (decoder->verbose > 1) block should be moved after the bitbuffer size check and preamble search. It looks like a Sparsnäs message has between 180 and 200 bits. That means that we only need to search for the preamble if we get that amount of bits from the demodulator. And the preamble search could be reduced to 0xAA, 0xD2, 0x01 for cases where parts of the preamble is missing.